src/python/pants/core/goals/test.py

```diff
@@ -869,6 +887,14 @@ async def _run_debug_tests(
     return Test(exit_code)
 
 
+def save_test_result_source_report_file(run_id: RunId, results: dict[str, str]) -> None:
```

Comment:
I think this becomes part of Pants' public API: if people start consuming this, we can't change it without breaking them. As such, maybe we could do one or more of:

1. call the option explicitly `experimental`
2. think about if we might want to evolve it in future, e.g. add more information about how a target was (or wasn't) run

For the first one, if we did that, I think we can land this with less "overhead": no need to be quite as precise if we're communicating to users that things may break.

For the second one, it seems like this currently maps `address -> status` directly, so we can't, add other metadata that ends up being useful (for instance, information about the execution environment or elapsed time or something). We might be able to solve this by having the metadata itself be a mapping `{"source": ...}`, e.g. for your example below:


```
$ cat test_result_source_report_runid0_2024_07_14_23_29_40.json | jq
{
  "timestamp": "2024_07_14_23_29_40",
  "run_id": 0,
  "sources": {
    "helloworld/greet/greeting_test.py:tests": {"source": "hit_locally"},
    "helloworld/translator/translator_test.py:tests": {"source": "hit_locally"}
  }
}

Reply:
Thanks very much, Huon! Every time I get a review from you, the PR becomes 10x better.

On the public API, isn't it that everything that can be imported from the `pants` wheel is public? I think that's the case. As we don't really have access modifiers, maybe just renaming the function to `_save_test_result_source_report_file` would suffice? :/ 

On the output data structure - the file being produced is named `test_result_source_report` so it contains only the sources information. If we don't extend the report, then it wouldn't make much sense to "stuff" more metadata into the file since it is only concerned with the sources.

If we think about future extensions, just brainstorming, what about this (note renamed file):

```
$ cat test_result_report_runid0_2024_07_14_23_29_40.json | jq

{
  "timestamp": "2024_07_14_23_29_40",
  "run_id": 0,
  "metadata": {
    "helloworld/greet/greeting_test.py:tests": {"source": "hit_locally", "re_run": True, attempts: 3},
    "helloworld/translator/translator_test.py:tests": {"source": "hit_locally", "re_run": False, attempts: 1}
  }
}
```

just an example of a few added pieces of metadata. This looks fairly extensible and doesn't require breaking any data structure contracts?

Reply:
> On the public API, isn't it that everything that can be imported from the pants wheel is public? I think that's the case. As we don't really have access modifiers, maybe just renaming the function to _save_test_result_source_report_file would suffice? :/


Ah sorry, I was a bit unclear, I was meaning the file format, rather than the Python function. I agree that pants isn't very explicit about its Python-level access control.

> This looks fairly extensible and doesn't require breaking any data structure contracts?

Looks good to me!